examples/dvbsrc.c

```diff
@@ -0,0 +1,271 @@
+    assert(uprobe_main != NULL);
+    uprobe_main = uprobe_uref_mgr_alloc(uprobe_main, uref_mgr);
+    assert(uprobe_main != NULL);
+    uprobe_main = uprobe_uclock_alloc(uprobe_main, uclock);
+    assert(uprobe_main != NULL);
+    uprobe_main = uprobe_ubuf_mem_pool_alloc(uprobe_main, umem_mgr,
+            UBUF_POOL_DEPTH, UBUF_SHARED_POOL_DEPTH);
+    assert(uprobe_main != NULL);
+    uprobe_main = uprobe_pthread_upump_mgr_alloc(uprobe_main);
+    assert(uprobe_main != NULL);
+    uref_mgr_release(uref_mgr);
+    uclock_release(uclock);
+    umem_mgr_release(umem_mgr);
+    uprobe_pthread_upump_mgr_set(uprobe_main, upump_mgr);
+
+    struct uprobe *uprobe_dejitter = uprobe_dejitter_alloc(uprobe_use(uprobe_main), true, 0);
+    assert(uprobe_dejitter != NULL);
+
+    struct upipe_mgr *src_xfer_mgr = upipe_pthread_xfer_mgr_alloc(255,
+            20, uprobe_use(uprobe_main), upump_ev_mgr_alloc_loop,
+            UPUMP_POOL, UPUMP_BLOCKER_POOL, NULL, NULL, NULL);
+    assert(src_xfer_mgr != NULL);
+    struct upipe_mgr *upipe_wsrc_mgr = upipe_wsrc_mgr_alloc(src_xfer_mgr);
+    assert(upipe_wsrc_mgr != NULL);
+    upipe_mgr_release(src_xfer_mgr);
+
+    uprobe_throw(uprobe_main, NULL, UPROBE_FREEZE_UPUMP_MGR);
+
+    /* dvb */
+    struct upipe_mgr *upipe_dvbsrc_mgr = upipe_dvbsrc_mgr_alloc();
+    struct upipe *upipe_src = upipe_void_alloc(upipe_dvbsrc_mgr,
+            uprobe_pfx_alloc(uprobe_use(uprobe_dejitter), loglevel , "dvbsrc"));
+    upipe_mgr_release(upipe_dvbsrc_mgr);
+    assert(upipe_src);
+    dvbsrc = upipe_src;
+
+    ubase_assert(upipe_set_uri(upipe_src, src));
+    ubase_assert(upipe_set_option(upipe_src, "sys", sys));
+    ubase_assert(upipe_set_option(upipe_src, "lnb", lnb));
+    ubase_assert(upipe_set_option(upipe_src, "frequency", freq));
+    ubase_assert(upipe_set_option(upipe_src, "symbol-rate", srate));
+    ubase_assert(upipe_set_option(upipe_src, "inner-fec", fec));
+    ubase_assert(upipe_set_option(upipe_src, "polarization", pol));
+    ubase_assert(upipe_attach_uclock(upipe_src));
+
+    uprobe_throw(uprobe_main, NULL, UPROBE_THAW_UPUMP_MGR);
+
+    /* deport to the source thread */
+    upipe_src = upipe_wsrc_alloc(upipe_wsrc_mgr,
+            uprobe_pfx_alloc(uprobe_use(uprobe_main),
+                UPROBE_LOG_VERBOSE, "wsrc"),
+            upipe_src,
+            uprobe_pfx_alloc(uprobe_use(uprobe_main),
+                UPROBE_LOG_VERBOSE, "wsrc_x"),
+            1000);
+
+    upipe_mgr_release(upipe_wsrc_mgr);
+
+    struct upipe_mgr *udpsink_mgr = upipe_udpsink_mgr_alloc();
+    struct upipe *udpsink = upipe_void_alloc_output(upipe_src, udpsink_mgr,
+            uprobe_pfx_alloc(uprobe_use(uprobe_main), loglevel, "udp sink"));
+    upipe_mgr_release(udpsink_mgr);
+    assert(udpsink);
+    ubase_assert(upipe_attach_uclock(udpsink));
+    upipe_release(udpsink);
+
+    ubase_assert(upipe_set_uri(udpsink, dst));
+
+    struct upump *timer = upump_alloc_timer(upump_mgr, stats_timer, NULL, NULL,
```

Comment:
Again, you are calling a control function on a pipe that runs in a different thread. It is not explicitly forbidden by Upipe specs, but if allowed on a specific control function the documentation should say so. Also, it would require libdvbv5 to be thread-safe, about which I cannot find any information. More importantly, what would happen if you modify the tuning parameters while the timer is running?

Reply:
Ah I removed the pthread but not the worker thread, that's fixed now.